server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2648,4 +2662,34 @@ public void afterRefresh(boolean didRefresh) throws IOException {
             refreshMetric.inc(System.nanoTime() - currentRefreshStartTime);
         }
     }
+
+    /** Rollback the current engine to the safe commit, the replay local translog up to the given {@code upToSeqNo} (inclusive) */
+    void resetEngineUpToSeqNo(long upToSeqNo) throws IOException {
+        assert getActiveOperationsCount() == 0 : "Ongoing writes [" + getActiveOperations() + "]";
+        assert SequenceNumbers.loadSeqNoInfoFromLuceneCommit(commitStats().getUserData().entrySet()).maxSeqNo == seqNoStats().getMaxSeqNo()
+            : "engine must be flushed before reset; max_seq_no[" + seqNoStats() + "] commit[" + commitStats().getUserData() + "]";
+        final Engine resettingEngine;
+        sync(); // persist the global checkpoint which will be used to trim unsafe commits
+        synchronized (mutex) {
+            final Engine currentEngine = getEngine();
+            final SeqNoStats seqNoStats = currentEngine.getSeqNoStats(getGlobalCheckpoint());
+            logger.info("resetting replica engine from max_seq_no [{}] to seq_no [{}] global checkpoint [{}]",
+                seqNoStats.getMaxSeqNo(), upToSeqNo, seqNoStats.getGlobalCheckpoint());
+            final TranslogStats translogStats = currentEngine.getTranslogStats();
+            final SearchOnlyEngine searchOnlyEngine = new SearchOnlyEngine(currentEngine.config(), seqNoStats, translogStats);
```

Comment:
I think it's ok to do this under the mutex since it's quick but why do we need to do this?